### PR TITLE
[PLAT-198] Implementation of SQL POSITION function

### PR DESCRIPTION
**Pull request by hakanmemisoglu**
* Adds support for POSITION(substring IN string [FROM position]) function

Checklist:
- [x] Labels and Milestone set

### CODE REVIEW

File: hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/string/HazelcastPositionFunction.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+                "POSITION",
+                SqlKind.OTHER_FUNCTION,
+                ReturnTypes.INTEGER_NULLABLE,
+                new ReplaceUnknownOperandTypeInference(SqlTypeName.VARCHAR),
+                SqlFunctionCategory.STRING
+        );
+    }
+
+    @Override
+    public String getSignatureTemplate(int operandCount) {
```
Comment (gierlachg) [inline hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/string/HazelcastPositionFunction.java:45]: I wonder why it's implemented in some functions while not in others. Is it actually used? Under what circumstances?

File: hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/string/HazelcastPositionFunction.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import org.apache.calcite.sql.type.SqlOperandCountRanges;
+import org.apache.calcite.sql.type.SqlTypeName;
+
+public class HazelcastPositionFunction extends HazelcastFunction {
+    public static final HazelcastPositionFunction INSTANCE = new HazelcastPositionFunction();
+
+    public HazelcastPositionFunction() {
+        super(
+                "POSITION",
+                SqlKind.OTHER_FUNCTION,
```
Comment (Fly-Style) [inline hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/string/HazelcastPositionFunction.java:37]:
```suggestion
                SqlKind.POSITION,
```

`POSITION` function is present in `SqlKind` enum :)

File: hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/string/PositionFunctionIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        check(sql("?", "?"), 2, "BC", "ABCD");
+    }
+
+    @Test
+    public void test_wrong_type() {
+        put("XYZ");
+
+        checkError(sql("123", "23"));
+        checkError(sql("TRUE", "FALSE"));
+        checkError(sql("?", "?"), 123, 23);
```
Comment (Fly-Style) [inline hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/string/PositionFunctionIntegrationTest.java:74]: Just asking for myself : did you test arguments with different types in your POSITION function?

**Review by Fly-Style — COMMENTED** [review]
Looks good, consider a few comments.

File: hazelcast/src/main/java/com/hazelcast/sql/impl/expression/string/StringFunctionUtils.java [hunk 4]
```diff
@@ -107,4 +107,8 @@ public static String asVarchar(Expression<?> expression, Row row, ExpressionEval
 
         return expression.getType().getConverter().asVarchar(res);
     }
+
+    public static Integer search(String text, String search, int start) {
```
[the remark this answers is not in the document]
Comment (hakanmemisoglu) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/expression/string/StringFunctionUtils.java:111]: I thought, we can implement functions like STRPOS, LOCATE etc. This method can be used in eval logic for these SQL function with an optional start argument.

File: hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/string/PositionFunctionIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        check(sql("?", "'BC'"), 2, "ABCD");
+        check(sql("'ABCD'", "?"), 2, "BC");
+        check(sql("?", "?"), 2, "BC", "ABCD");
+    }
+
+    @Test
+    public void test_wrong_type() {
+        put("XYZ");
+
+        checkError(sql("123", "23"));
```
Comment (viliam-durina) [inline hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/string/PositionFunctionIntegrationTest.java:72]: We should also assert the error message.

File: hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/string/HazelcastPositionFunction.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+                ReturnTypes.INTEGER_NULLABLE,
+                new ReplaceUnknownOperandTypeInference(SqlTypeName.VARCHAR),
+                SqlFunctionCategory.STRING
```
[the remark this answers is not in the document]
Comment (hakanmemisoglu) [reply hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/string/HazelcastPositionFunction.java:40]: It looks like Calcite uses return type as function category. I could not find any Calcite logic affected by this decision.
Anyways, I will change it to Numeric.

File: hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/string/HazelcastPositionFunction.java (hunk 1 above)
Comment (hakanmemisoglu) [reply hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/string/HazelcastPositionFunction.java:45]: I have put a breakpoint but it didn't get triggered in test cases.
After looking the code, `calcite.SqlCallBinding.newValidationSignatureError()` seems the only place indirectly reaches this call.
We have `HazelcastCallBinding` which overrides it, and Calcite code calls the overridden method.

Comment (gierlachg) [reply hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/string/HazelcastPositionFunction.java:52]: So, if I understand correctly it's never used. Maybe we should simply remove it (and all other overrides)?

Comment (hakanmemisoglu) [reply hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/string/HazelcastPositionFunction.java:52]: OK, I will remove `getSignatureTemplate`. 
I have added `unparse` override and what I have seen from the code, it is used in several places.

File: hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/string/PositionFunctionIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+    public void test() {
+        put("XYZ");
+
+        check(sql("'ABCD'", "'BC'"), 2);
+        check(sql("'ABCD'", "''"), 1);
+        check(sql("'ABCD'", "'NO_MATCH'"), 0);
+
+
+        // Test with start argument, slide the search index to right.
+        // Valid start indices are between 0 and 4.
```
Comment (viliam-durina) [inline hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/string/PositionFunctionIntegrationTest.java:63]:
```suggestion
        // Valid start indices are between 1 and 4.
```

**Review by viliam-durina — CHANGES_REQUESTED** [review]

File: hazelcast/src/main/java/com/hazelcast/sql/impl/expression/string/PositionFunction.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        if (text == null) {
+            return null;
+        }
+        String search = StringFunctionUtils.asVarchar(operand2, row, context);
+        if (search == null) {
+            return null;
+        }
+
+        // Start argument is not given, then search it from the beginning.
+        if (operand3 == null) {
```
[the remark this answers is not in the document]
Comment (viliam-durina) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/expression/string/PositionFunction.java:67]: I was wrong, `operand3` is null if there's no 3rd argument and it's `ConstantExpression{null}`, when there is 3rd argument, but it's null.

**Review by viliam-durina — APPROVED** [review]

**Review by gierlachg — APPROVED** [review]

**Comment by hakanmemisoglu** [comment]
run-lab-run

**Comment by viliam-durina** [comment]
run-lab-run

**Pull request merged**